onnxruntime/core/graph/constant_folding.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/graph/constant_folding.h"
+#include "core/graph/graph_utils.h"
+#include "core/session/inference_session.h"
+#include "core/graph/model.h"
+#include "core/framework/ml_value.h"
+
+namespace onnxruntime {
+
+Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified) {
+  // TODO Can we reuse the model across calls of the constant folding rule?
+  // TODO Instead of computing one node at a time, we can traverse the whole graph and
+  // find constant parts of it. Then we can create bigger subgraphs to compute directly.
+  // This will be a useful optimization if we expect big subgraphs to be constant, and
+  // should be done through a Transformer, not a RewriteRule.
+  auto p_model = std::make_unique<onnxruntime::Model>("ConstantFoldingModel", false, ModelMetaData(),
+                                                      IOnnxRuntimeOpSchemaRegistryList({graph.GetSchemaRegistry()}),
+                                                      graph.DomainToVersionMap());
+  Graph& subgraph = p_model->MainGraph();
+
+  std::vector<onnxruntime::NodeIndex> subgraph_nodes;
+  subgraph_nodes.push_back(node.Index());
+
+  // Build the subgraph.
+  graph_edit_utils::BuildSubgraph(graph, subgraph_nodes, subgraph);
+
+  SessionOptions so;
+  so.session_logid = "ConstantFoldingSession";
+  // Disable default graph transformers for the constant node (can lead to infinite recursion otherwise).
+  so.enable_default_transformers = false;
+  InferenceSession session_object{so};
+  // TODO Make it possible to pass the model directly to the session instead of having to dump it to a stream.
+  std::stringstream model_istream;
```

Comment:
Would be good to address this and cleanup the 4 cut-and-pasted versions of InferenceSession::Impl::Load. 

Maybe a generic Load method that takes a std::function<shared_ptr<Model>(void)> could be added. The existing versions could call that with a lambda to abstract out their model loading logic, and this code could also call that with a lambda that just returns p_model.

If you don't have time please add a Task and I'll take care of it.

Reply:
Yeah, that would be nice. If you can take care of it, it would be great.

Reply:
We no longer need to create a model, so we don't need the serialization.